File: Code Python/04. Denoising Data/04.2 Denoising Data with FFT/DenoisingImageWithFFT2.py
import matplotlib.pyplot as plt
import numpy as np
from matplotlib.image import imread

## Import Data White Noise
imageimp = imread('Einstein.jpg')
image = np.double(imageimp)

noise = np.random.randn(3668, 2945)*100

noisyimage = image + noise

## Compute Fourier Coefficient matrix

FC = np.fft.fft2(noisyimage)

## Filter Noise Frequencies
# only pass low frequencies (middle)

FCshifted = np.fft.fftshift(FC)
m = round(3668/2)
n = round(2945/2)

a = 190
FCclean = np.zeros((3668, 2945), dtype=complex)
FCclean[m-a:m+a, n-a:n+a] = FCshifted[m-a:m+a, n-a:n+a]


## Reconstruct denoised image

FCclean = np.fft.fftshift(FCclean)
Imagedenoised = np.fft.ifft2(FCclean)

plt.figure(figsize=(20, 10))
plt.subplot(2,3,1)
plt.axis('off')
plt.imshow(noisyimage, cmap='gray')
plt.title('Noisy image')
plt.subplot(2,3,2)
plt.axis('off')
plt.imshow(abs(Imagedenoised), cmap='gray')
plt.title('Denoised image')
plt.axis('off')
plt.subplot(2,3,3)
plt.imshow(image, cmap='gray')
plt.title('Original')
plt.axis('off')

FClog = np.log(abs(np.fft.fftshift(FC))+1)
plt.subplot(2,3,4)
plt.imshow(FClog, cmap='gray')
plt.axis('off')

FCcleanlog = np.log(abs(np.fft.fftshift(FCclean))+1)
plt.subplot(2,3,5)
plt.imshow(FCcleanlog, cmap='gray')
plt.axis('off')

FCcleanlog = np.log(abs(np.fft.fftshift(np.fft.fft2(image)))+1)
plt.subplot(2,3,6)
plt.imshow(FCcleanlog, cmap='gray')
plt.axis('off')
plt.show()

## Import Data High Frequent Noise

Import = imread('NoisyMoonLanding.png')
Image = np.double(Import)

## Plot Noisy Picture

plt.figure()
plt.imshow(Image, cmap='gray')
plt.axis('off')
plt.show()

## Compute Fourier Coefficient matrix

FC = np.fft.fft2(Image)

## Filter Noise Frequencies
# Noise Frequnences should be the dominant frequences of the picture.

FClog = np.log(abs(np.fft.fftshift(FC))+1)
plt.figure()
plt.subplot(2,1,1)
plt.imshow(FClog, cmap='gray')
plt.axis('off')

# only pass low frequencies (middle)

FCshifted = np.fft.fftshift(FC)
m = round(474/2)
n = round(630/2)

a = 44
FCclean = np.zeros((474, 630), dtype=complex)
FCclean[m-a:m+a, n-a:n+a] = FCshifted[m-a:m+a, n-a:n+a]

FCcleanlog = np.log(abs(FCclean)+1)
plt.subplot(2,1,2)
plt.imshow(FCcleanlog, cmap='gray')
plt.axis('off')
plt.show()

## Reconstruct denoised image

FCclean = np.fft.fftshift(FCclean)
Imagedenoised = np.fft.ifft2(FCclean)

plt.figure(figsize=(20,10))
plt.subplot(1,2,1)
plt.imshow(Image, cmap='gray')
plt.title('Noisy image')
plt.axis('off')
plt.subplot(1,2,2)
plt.imshow(abs(Imagedenoised), cmap='gray')
plt.title('Denoised image')
plt.axis('off')
plt.show()

## Filtering Noise special Frequencies

FClog = np.log(abs(np.fft.fftshift(FC))+1)
X, Y = np.meshgrid(np.arange(1, np.shape(FClog)[1]+1), np.arange(1, np.shape(FClog)[0]+1))
figs = plt.figure(figsize=(10, 10))
sp = figs.add_subplot(211, projection='3d')
mappable = sp.plot_surface(X, Y, FClog, cmap='jet')

ind = FClog < 3.9
a = 44
b = 110
indinner = FClog < 4.8
FCclean2 = FCshifted*ind
FCclean2inner = FCshifted*indinner
FCclean2[m-b:m+b, n-b:n+b] = FCclean2inner[m-b:m+b, n-b:n+b]
FCclean2[m-a:m+a, n-a:n+a] = FCshifted[m-a:m+a, n-a:n+a]

FCclean2log = np.log(abs(FCclean2)+1)
plt.subplot(2,1,2)
plt.imshow(FCclean2log, cmap='gray')
plt.axis('off')
plt.show()


FCclean = np.fft.fftshift(FCclean2)
Imagedenoised = np.fft.ifft2(FCclean2)

plt.figure(figsize=(20,10))
plt.subplot(1,2,1)
plt.imshow(Image, cmap='gray')
plt.axis('off')
plt.title('Noisy image')
plt.subplot(1,2,2)
plt.imshow(abs(Imagedenoised),cmap='gray')
plt.axis('off')
plt.title('Denoised image')
plt.show()

File: Code Matlab/04. Denoising Data/04.2 Denoising Data with FFT/DenoisingImageWithFFT2.m
clc, clear, close all

%% Import Data White Noise
imageimp = imread('Einstein.jpg');
image = double(imageimp);

noise = randn(size(image))*100;

noisyimage = image + noise;

%% Compute Fourier Coefficient matrix

FC = fft2(noisyimage);

%% Filter Noise Frequencies
% only pass low frequencies (middle)

FCshifted = fftshift(FC);
m = round(size(FCshifted,1)/2);
n = round(size(FCshifted,2)/2);

a = 190;
FCclean = zeros(size(FCshifted));
FCclean(m-a:m+a,n-a:n+a)=FCshifted(m-a:m+a,n-a:n+a);


%% Reconstruct denoised image

FCclean = fftshift(FCclean);
Imagedenoised = ifft2(FCclean);

figure('Name','Denoised Picture');
subplot(2,3,1)
imshow(uint8(noisyimage));
title('Noisy image')
subplot(2,3,2)
imshow(uint8(abs(Imagedenoised)));
title('Denoised image')
subplot(2,3,3)
imshow(uint8(image));
title('Original')

FClog = log(abs(fftshift(FC))+1);
subplot(2,3,4)
imshow(mat2gray(FClog));

FCcleanlog = log(abs(fftshift(FCclean))+1);
subplot(2,3,5)
imshow(mat2gray(FCcleanlog));

FCcleanlog = log(abs(fftshift(fft2(image)))+1);
subplot(2,3,6)
imshow(mat2gray(FCcleanlog));

%% Import Data High Frequent Noise

Import = imread('NoisyMoonLanding.png');
Image = double(Import);


%% Compute Fourier Coefficient matrix

FC = fft2(Image);

%% Filter Noise Frequencies

FClog = log(abs(fftshift(FC))+1);
figure('Name','Fourier coefficients')
subplot(2,2,3)
imshow(mat2gray(FClog));

% only pass low frequencies (middle)

FCshifted = fftshift(FC);
m = size(FCshifted,1);
n = size(FCshifted,2);

a = 44;
FCclean = zeros(size(FCshifted));
FCclean(m/2-a:m/2+a,n/2-a:n/2+a)=FCshifted(m/2-a:m/2+a,n/2-a:n/2+a);

FCcleanlog = log(abs(FCclean)+1);
subplot(2,2,4)
imshow(mat2gray(FCcleanlog));


%% Reconstruct denoised image

FCclean = fftshift(FCclean);
Imagedenoised = ifft2(FCclean);

subplot(2,2,1)
imshow(uint8(Image));
title('noisy image')
subplot(2,2,2)
imshow(uint8(abs(Imagedenoised)));
title('denoised image')

%% Filtering Noise special Frequencies

FClog = log(abs(fftshift(FC))+1);
figure('Name','Fourier coefficients')
subplot(1,2,1)
hold on
surf(FClog, 'EdgeColor','interp'),colormap('jet');
surf(ones(size(FClog))*9.9,'FaceColor','b','EdgeColor','b');

a = 44;
b = 110;

s=ones(size(FClog))*11;
surf(n/2-b:n/2+b,m/2-b:m/2+b,s(m/2-b:m/2+b,n/2-b:n/2+b),'FaceColor','k');
legend('Fourier coefficients','Filtering layer 1','Inner filtering layer')

ind = FClog < 9.9;
indinner = FClog < 11;
FCclean2 = FCshifted.*ind;
FCclean2inner = FCshifted.*indinner;
FCclean2(m/2-b:m/2+b,n/2-b:n/2+b)=FCclean2inner(m/2-b:m/2+b,n/2-b:n/2+b);
FCclean2(m/2-a:m/2+a,n/2-a:n/2+a)=FCshifted(m/2-a:m/2+a,n/2-a:n/2+a);

FCclean2log = log(abs(FCclean2)+1);
subplot(1,2,2)
imshow(mat2gray(FCclean2log));


FCclean = fftshift(FCclean2);
Imagedenoised = ifft2(FCclean2);

figure('Name','Denoised Picture');
subplot(1,2,1)
imshow(uint8(Image));
title('noisy image')
subplot(1,2,2)
imshow(uint8(abs(Imagedenoised)));
title('denoised image')
